src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.State.cs

```diff
@@ -0,0 +1,297 @@
+                }
+
+                var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+                var solution = document.Project.Solution;
+                var documents = documentIds.Select(id => solution.GetDocument(id)).ToImmutableArray();
+
+                var result = await TryGetDefaultNamespaceAndNamespaceDeclarationAsync(
+                    documents, service, syntaxFacts, textSpan, cancellationToken).ConfigureAwait(false);
+
+                if (!result.shouldTriggerRefactoring)
+                {
+                    return null;
+                }
+
+                var defaultNamespace = result.defaultNamespace;
+                var declaredNamespace = result.declaredNamespace;
+
+                // Namespace can't be renamed if we can't construct a valid qualified identifier from folder names.
+                // In this case, we might still be able to provide refactoring to move file to new location.
+                var namespaceFromFolders = TryBuildNamespaceFromFolders(service, document.Folders, syntaxFacts);
+                var targetNamespace = namespaceFromFolders == null ? null : ConcatNamespace(defaultNamespace, namespaceFromFolders);      
+
+                // No action required if namespace already matches folders.
+                if (syntaxFacts.StringComparer.Equals(targetNamespace, declaredNamespace))
+                {
+                    return null;
+                }
+
+                // Only provide "move file" action if root namespace contains declared namespace.
+                // It makes no sense to match folder hierarchy with namespace if it's not rooted at the root namespace of the project. 
+                var relativeNamespace = GetRelativeNamespace(defaultNamespace, declaredNamespace, syntaxFacts);                                                                           
+                                                                                                                          
+                return new State(solution, document.Id, documentIds, defaultNamespace, targetNamespace, declaredNamespace, relativeNamespace);
+            }
+
+            /// <summary>
+            /// Create a qualified identifier as the suffix of namespace based on a list of folder names.
+            /// </summary>
+            private static string TryBuildNamespaceFromFolders(
+                AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax> serivice, 
+                IEnumerable<string> folders, 
+                ISyntaxFactsService syntaxFacts)
+            {
+                var isFirst = true;
+                var builder = PooledStringBuilder.GetInstance();
+
+                // In case a folder name is a qualified identifier, e.g. `A.B`, we need to make
+                // sure every part is a valid identifier and escape it properly.
+                foreach (var part in folders.SelectMany(folder => folder.Split(new[] { '.' })))
```

Comment:
this can all be simplified without the need for a for-loop or keepign track of first or not.  just do:

var parts = folders.SelectMany(folder => folder.Split(new[] { '.' }).SelectAsArray(service.EscapeIdentifier);
if (!parts.All(syntaxFacts.IsValidIdentifier) { return null; }
return string.join(".", parts);

Reply:
Neat 👍